specs/protocol/safer-safes.md

```diff
@@ -0,0 +1,559 @@
+
+#### `clearTimelockGuard`
+
+Remove the timelock guard configuration by a previously enabled `safe`.
+
+- MUST revert if the contract is enabled as a guard for the `safe`.
+- MUST revert if the contract is not configured for the `safe`.
+- MUST erase the existing `timelock_delay` data related to the calling `safe`.
+- MUST emit a `GuardCleared` event.
+
+#### `scheduleTransaction`
+
+Called by anyone using signatures from Safe owners, registers a transaction in the TimelockGuard for execution after
+the `timelock_delay`.
+
+- MUST revert if the contract is not enabled as a guard for the `safe`.
+- MUST revert if the contract is not configured for the `safe`.
+- MUST take the same parameters as `execTransaction`.
+- MUST revert if an identical transaction has already been scheduled.
+- MUST revert if an identical transaction was cancelled.
+- MUST set `execution_time(safe, txHash)` to `block.timestamp + timelock_delay(safe)`.
+- MUST emit a `TransactionScheduled` event, with at least `safe` and relevant transaction data.
+
+#### `cancelTransaction`
+
+Makes a scheduled transaction not executable. To do so, it builds a no-op transaction at the same nonce as the
+scheduled transaction, and verifies that the supplied signatures for such a no-op transaction are valid and amount to
+`cancellation_threshold(safe)`. If successful, it increases the `cancellation_threshold(safe)` by 1.
+
+- MUST revert if the contract is not enabled as a guard for the `safe`.
+- MUST revert if the contract is not configured for the `safe`.
+- MUST revert if `sum(rejecting_owners(safe, tx)) < cancellation_threshold(safe)`.
+- MUST emit a `TransactionCancelled` event, with at least `safe` and a transaction identifier.
+
+#### `checkTransaction`
+
+Can be called by anyone. It will be called by the `safe` in `execTransaction` if the contract is enabled as a guard.
+It verifies if a given transaction was scheduled and the delay period has passed, and reverts if not.
+
+- MUST revert if the contract is not enabled as a guard for the `safe`.
+- MUST revert if the contract is not configured for the `safe`.
+- MUST take the exact parameters from the `ITransactionGuard.checkTransaction` interface.
+- MUST revert if `execution_time(safe, tx) < block.timestamp`.
+- MUST revert if the scheduled transaction was cancelled.
+
+#### `checkAfterExecution`
+
+Can be called by anyone. It will be called by the `safe` in `execTransaction` if the contract is enabled as a guard.
+It updates state after the transaction has been executed.
+
+- MUST take the exact parameters from the `ITransactionGuard.checkTransaction` interface.
+- MUST revert if the contract is not enabled as a guard for the `safe`. // TODO: I'm not sure if we want to ever revert here
+- MUST revert if the contract is not configured for the `safe`. // TODO: I'm not sure if we want to ever revert here
+- MUST set `cancellation_threshold(safe)` to 1.
```

Comment:
IMO we shold do this in the `checkTransaction()` function. It's simpler and follows the checks, effects and interactions pattern.

Reply:
True, removed the whole after-execution function from the spec.